lib/logstash/inputs/kafka.rb

```diff
@@ -20,168 +21,89 @@
-        :rebalance_backoff_ms => @rebalance_backoff_ms,
-        :consumer_timeout_ms => @consumer_timeout_ms,
-        :consumer_restart_on_error => @consumer_restart_on_error,
-        :consumer_restart_sleep_ms => @consumer_restart_sleep_ms,
-        :consumer_id => @consumer_id,
-        :fetch_message_max_bytes => @fetch_message_max_bytes,
-        :allow_topics => @white_list,
-        :filter_topics => @black_list,
-        :value_decoder_class => @decoder_class,
-        :key_decoder_class => @key_decoder_class
-    }
-    if @reset_beginning
-      options[:reset_beginning] = 'from-beginning'
-    end # if :reset_beginning
-    topic_or_filter = [@topic_id, @white_list, @black_list].compact
-    if topic_or_filter.count == 0
-      raise LogStash::ConfigurationError, 'topic_id, white_list or black_list required.'
-    elsif topic_or_filter.count > 1
-      raise LogStash::ConfigurationError, 'Invalid combination of topic_id, white_list or black_list. Use only one.'
-    end
-    @kafka_client_queue = SizedQueue.new(@queue_size)
-    @consumer_group = create_consumer_group(options)
-    @logger.info('Registering kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
+    @message_count = 0
+    @runner_threads = []
   end # def register
 
   public
   def run(logstash_queue)
-    # noinspection JRubyStringImportInspection
-    java_import 'kafka.common.ConsumerRebalanceFailedException'
-    @logger.info('Running kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
-    begin
-      @consumer_group.run(@consumer_threads,@kafka_client_queue)
-
-      while !stop?
-        if !@kafka_client_queue.empty?
-          event = @kafka_client_queue.pop
-          queue_event(event, logstash_queue)
-        end
-      end
-
-      until @kafka_client_queue.empty?
-        queue_event(@kafka_client_queue.pop,logstash_queue)
-      end
-
-      @logger.info('Done running kafka input')
-    rescue => e
-      @logger.warn('kafka client threw exception, restarting',
-                   :exception => e)
-      Stud.stoppable_sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000) { stop? }
-      retry if !stop?
-    end
+    @runner_consumers = num_threads.times.map { || new_consumer }
+    @runner_threads = @runner_consumers.map { |consumer| thread_runner(logstash_queue, consumer) }
+    @runner_threads.each { |t| t.join }
   end # def run
 
   public
   def stop
-    @consumer_group.shutdown if @consumer_group.running?
+    @runner_consumers.each { |c| c.wakeup }
   end
 
   private
-  def create_consumer_group(options)
-    Kafka::Group.new(options)
+  def thread_runner(logstash_queue, consumer)
+    Thread.new do
+      begin
+        consumer.subscribe(topics);
+        while !stop?
+          records = consumer.poll(timeout_ms);
+          for record in records do
+            break if !max_messages.nil? && @message_count >= max_messages
+            @message_count = @message_count.next
```

Comment:
I'd remove this feature unless someone explicitly asks for consuming `n` events and shutting down. WDYT?